Remove required positional arguments from `run-external` and `exec`

132ikl opened the pull request:
<!--
if this PR closes one or more issues, you can automatically link the PR with
them by using one of the [*linking keywords*](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword), e.g.
- this PR should close #xxxx
- fixes #xxxx

you can also mention related issues, PRs or discussions!
-->

# Description
<!--
Thank you for improving Nushell. Please, check our [contributing guide](../CONTRIBUTING.md) and talk to the core team before making major changes.

Description of your pull request goes here. **Provide examples and/or screenshots** if your changes affect the user experience.
-->

This PR removes the required positional argument from `run-external` and `exec` in favor of the rest arguments, meaning lists of external commands can be spread directly into `run-external` and `exec`. This does have the drawback of making calling `run-external` and `exec` with no arguments a run-time error rather than a parse error, but I don't imagine that is an issue.

Before (for both `run-external` and `exec`):
```nushell
run-external
# => Error: nu::parser::missing_positional
# => 
# =>   × Missing required positional argument.
# =>    ╭─[entry #9:1:13]
# =>  1 │ run-external
# =>    ╰────
# =>   help: Usage: run-external <command> ...(args) . Use `--help` for more
# =>         information.

let command = ["cat" "hello.txt"]
run-external ...$command
# => Error: nu::parser::missing_positional
# => 
# =>   × Missing required positional argument.
# =>    ╭─[entry #11:1:14]
# =>  1 │ run-external ...$command
# =>    ·              ▲
# =>    ·              ╰── missing command
# =>    ╰────
# =>   help: Usage: run-external <command> ...(args) . Use `--help` for more
# =>         information.
run-external ($command | first) ...($command | skip 1)
# => hello world!
```

After (for both `run-external` and `exec`):
```nushell
run-external
# => Error: nu::shell::missing_parameter
# => 
# =>   × Missing parameter: no command given.
# =>    ╭─[entry #2:1:1]
# =>  1 │ run-external
# =>    · ──────┬─────
# =>    ·       ╰── missing parameter: no command given
# =>    ╰────
# => 

let command = ["cat" "hello.txt"]
run-external ...$command
# => hello world!
```



# User-Facing Changes
<!-- List of all changes that impact the user experience here. This helps us keep track of breaking changes. -->

Lists can now be spread directly into `run-external` and `exec`:

```nushell
let command = [cat hello.txt]
run-external ...$command
# => hello world!
``` 

# Tests + Formatting
<!--
Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
- `cargo run -- -c "use toolkit.nu; toolkit test stdlib"` to run the tests for the standard library

> **Note**
> from `nushell` you can also use the `toolkit` as follows
> ```bash
> use toolkit.nu  # or use an `env_change` hook to activate it automatically
> toolkit check pr
> ```
-->

- :green_circle: `toolkit fmt`
- :green_circle: `toolkit clippy`
- :green_circle: `toolkit test`
- :green_circle: `toolkit test stdlib`

# After Submitting
<!-- If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date. -->
N/A

132ikl commented:
Actually wait, this isn't a breaking change. I'm not sure why I thought it was. The old format still works in the after the PR, since it's still just rest arguments.

```nushell
let command = ["cat" "hello.txt"]
run-external ...$command
# => hello world!
run-external ($command | first) ...($command | skip 1)
# => hello world!
```

WindSoilder commented on crates/nu-command/src/system/run_external.rs:
```diff
@@ -34,15 +34,10 @@ impl Command for External {
-                SyntaxShape::OneOf(vec![SyntaxShape::GlobPattern, SyntaxShape::Any]),
-                "Arguments for external command.",
+                "command",
+                SyntaxShape::OneOf(vec![SyntaxShape::Any, SyntaxShape::GlobPattern]),
```
I think the order of the `vector` is important.  If we put `SyntaxShape::Any` first, it will never have a chance to fall to `SyntaxShape::GlobPattern`.
Can you explain why you want to change the order?

132ikl replied:
oh, I see, thanks for catching this. I switched it because the syntax highlighting looked weird, and according to the AST it was trying to interpret everything as glob pattern literals(?). I thought it was due to my changes, but it looks like it's actually already like this on main.

Here's what it looks like on `main`:
![image](https://github.com/user-attachments/assets/da861ebb-7ed0-446a-a951-a71ae940d1e1)

Here's what it looks like after this PR:

![image](https://github.com/user-attachments/assets/b8e6d6ef-a861-424d-92be-6b76387e2446)

..except this is also what `ls` looks like, so I'm not sure what I was thinking.

It does seem to break the globbing. It looks like the tests didn't catch this because they all use the `^command` syntax instead of `run-external`. I'll revert this, and add some tests to test with `run-external` specifically.

132ikl commented:
I added test permutations in `tests/commands/run_external.rs` to run each test with just the binary name, with `^`, and with `run-external`. There were a couple special cases where not all permutations could be used (notably the parser seems to handle GlobPatterns strangely, which only affects run-external), and I left a comment to explain the issue for each of these. I also verified that the updated tests caught the bug I accidentally introduced by switching `SyntaxShape::GlobPattern` and `SyntaxShape::Any` in `run-external`'s signature (and reverted that change).

Additionally, I pulled in a new dependency to help with these tests, [`rstest_reuse`](https://github.com/la10736/rstest/tree/master/rstest_reuse). This crate is actively maintained directly in the main `rstest` repo, and it didn't pull any further dependencies which we don't already have, so hopefully this shouldn't be a major concern.

WindSoilder commented on crates/nu-command/src/system/exec.rs:
```diff
@@ -11,7 +13,11 @@ impl Command for Exec {
-            .required("command", SyntaxShape::String, "The command to execute.")
+            .rest(
+                "command",
+                SyntaxShape::OneOf(vec![SyntaxShape::Any, SyntaxShape::GlobPattern]),
```
I think it should be changed to `SyntaxShape::OneOf(vec![SyntaxShape::GlobPattern, SyntaxShape::Any])` too.

132ikl replied:
oops, thanks!

132ikl commented on crates/nu-command/src/system/run_external.rs:
```diff
@@ -34,15 +34,10 @@ impl Command for External {
             .input_output_types(vec![(Type::Any, Type::Any)])
-            .required(
-                "command",
-                SyntaxShape::OneOf(vec![SyntaxShape::GlobPattern, SyntaxShape::String]),
```
I think it should just turn these parse errors into runtime errors, but I'll tested it out

132ikl replied:
It seems like putting literals for most types in the command position just attempt to run that literal as a command anyways, which is fine. If it can't be coerced into a string it will simply error at runtime. The only difference I found was bools:

Before:
```
$ run-external false
Error: nu::parser::parse_mismatch_with_full_string_msg

  × Parse mismatch during operation.
   ╭─[entry #18:1:14]
 1 │ run-external false
   ·              ──┬──
   ·                ╰── expected one of a list of accepted shapes: [GlobPattern, String]
   ╰────
```

After:

```
run-external false
```
(runs `false` executable)

Seems like the after behavior is preferable here.

WindSoilder reviewed (approved):
Thanks!  LGTM
